[ZEPPELIN-1086] Auto Completion for Interpreter.

astroshim opened the pull request:
### What is this PR for?
This PR is to support auto-completion for interpreter.


### What type of PR is it?
Improvement 


### What is the Jira issue?
https://issues.apache.org/jira/browse/ZEPPELIN-1086


### How should this be tested?
Type ```% + 'Ctrl+.' ``` in your paragraph.
please refer to screen shot.


### Screenshots (if appropriate)
![o](https://cloud.githubusercontent.com/assets/3348133/16543637/a84a0f70-4119-11e6-9790-d5b3913fe536.gif)


### Questions:
* Does the licenses files need update? no
* Is there breaking changes for older versions? no
* Does this needs documentation? no

astroshim commented:
re-trigger CI

AhyoungRyu commented:
@astroshim Awesome! :+1:  
I'm not sure where can be the proper place, it would be better users can know this feature's existence since it's really useful. How about adding this info in [here](https://github.com/apache/zeppelin/blob/master/zeppelin-web/src/components/modal-shortcut/modal-shortcut.html#L211)? This is just my personal opinion :)

astroshim commented:
@AhyoungRyu Thank you for your review!
I think it's better to make zeppelin manual on zeppelin homepage than adding shortcut info page. What do you think?

AhyoungRyu commented:
@astroshim Oh sorry for my late response. Yeah that would be better 👍

astroshim commented:
@bzz @Leemoonsoo please review.

Leemoonsoo commented on zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Paragraph.java:
```diff
@@ -202,11 +201,26 @@ public Interpreter getCurrentRepl() {
   public List<InterpreterCompletion> completion(String buffer, int cursor) {
+    String lines[] = buffer.split(System.getProperty("line.separator"));
+    if (lines.length > 0
+      && (lines[0].startsWith("%") && cursor <= lines[0].trim().length())) {
```
Code text can follow after interpreter selection without line.separator. For example

```sql
%sql select * from table
```
In this case, `getInterpreterCompletion()` will be invoked even if cursor is in somewhere `select * from table`. isn't it?

astroshim replied:
You're right. I missed that point. Thank you pointing out.

felixcheung replied:
space or new line could be separator in this case?

astroshim replied:
yes. It should be separated space or new line.

astroshim commented:
@Leemoonsoo As you can see the following screenshot, I fixed bug and changed completion results.

![o](https://cloud.githubusercontent.com/assets/3348133/16755681/2918e9b0-483a-11e6-96cf-13c3438d70b5.gif)

bzz commented:
@astroshim looks awesome, thank you for implementing!

How do you think, would it improve user experience with this feature if we could remove from suggestion list things, that are not interpreters? Like function names, etc. And how hard would that be to implement?

astroshim commented:
@bzz Thank you for reviewing. 
You mean that blocking the keyword and local meta values?

bzz commented:
What I mean is - as a user, when I auto-complete the interpreter name, I would expect a suggestion list that contains only interpreter names and nothing else.

Right now list of suggestion includes things like "NoSuchMethod" and `AbstractMethodError`.

If it is hard to remove such things from the suggestions list in this case - let's keep them. But if that is something we could implement in this PR - that would be nice.

How do you think?

astroshim commented:
@bzz I totally agree with you so I did try to remove things you mentioned but it's not easy.
Let me try one more. Thanks!

bzz commented:
Got it! Thanks a lot for looking into it.

Let me know when you think its ready and I will try again. You did great job already, making it auto-complete names so in case it's too hard - we could merge as it is now, its already very useful!

astroshim commented:
@bzz Thank you kindly for telling like that. 
What do you think about tagging meta value than removing keywords like following image? 
![image](https://cloud.githubusercontent.com/assets/3348133/16870311/eb05d992-4abb-11e6-8172-99c90196a5c9.png)

I think it looks more nicer than removing keywords.
If it is okay, let me make it at the another PR.

bzz commented:
Sounds as a great plan to me!

Merging this one then, if there is no further discussion

astroshim commented:
Ping

bzz commented:
Sorry, it took me a while to get back. Done!

minahlee commented:
@astroshim this PR causes compile error in current master. Do you mind to create another PR to fix it?

astroshim commented:
@minahlee Thank you for addressing that. I'll take a look.

astroshim commented:
@minahlee Sorry for bothering you. I just made PR for fixing compile error.
